metrics-for-spark-clusters.md

```diff
@@ -0,0 +1,248 @@
+
+To effectively enable a Graphite based metrics solution, the pod spec that
+Oshinko generates for Spark clusters will need to be changed to include the
+two Graphite based contianers and a service will need to be created to allow
+access to the metrics data. Images will also need to be generated for these
+new processes, and the Spark images will need to have their configuration
+files changed to allow metrics reporting.
+
+**Example 1, sample metrics topics**
+
+```
+"jvm.PS-MarkSweep.count", 
+"jvm.PS-MarkSweep.time", 
+"jvm.PS-Scavenge.count", 
+"jvm.PS-Scavenge.time", 
+"jvm.heap.committed", 
+"jvm.heap.init", 
+"jvm.heap.max", 
+"jvm.heap.usage", 
+"jvm.heap.used", 
+"jvm.non-heap.committed", 
+"jvm.non-heap.init", 
+"jvm.non-heap.max", 
+"jvm.non-heap.usage", 
+"jvm.non-heap.used", 
+"jvm.pools.Code-Cache.committed", 
+"jvm.pools.Code-Cache.init", 
+"jvm.pools.Code-Cache.max", 
+"jvm.pools.Code-Cache.usage", 
+"jvm.pools.Code-Cache.used", 
+"jvm.pools.Compressed-Class-Space.committed", 
+"jvm.pools.Compressed-Class-Space.init", 
+"jvm.pools.Compressed-Class-Space.max", 
+"jvm.pools.Compressed-Class-Space.usage", 
+"jvm.pools.Compressed-Class-Space.used", 
+"jvm.pools.Metaspace.committed", 
+"jvm.pools.Metaspace.init", 
+"jvm.pools.Metaspace.max", 
+"jvm.pools.Metaspace.usage", 
+"jvm.pools.Metaspace.used", 
+"jvm.pools.PS-Eden-Space.committed", 
+"jvm.pools.PS-Eden-Space.init", 
+"jvm.pools.PS-Eden-Space.max", 
+"jvm.pools.PS-Eden-Space.usage", 
+"jvm.pools.PS-Eden-Space.used", 
+"jvm.pools.PS-Old-Gen.committed", 
+"jvm.pools.PS-Old-Gen.init", 
+"jvm.pools.PS-Old-Gen.max", 
+"jvm.pools.PS-Old-Gen.usage", 
+"jvm.pools.PS-Old-Gen.used", 
+"jvm.pools.PS-Survivor-Space.committed", 
+"jvm.pools.PS-Survivor-Space.init", 
+"jvm.pools.PS-Survivor-Space.max", 
+"jvm.pools.PS-Survivor-Space.usage", 
+"jvm.pools.PS-Survivor-Space.used", 
+"jvm.total.committed", 
+"jvm.total.init", 
+"jvm.total.max", 
+"jvm.total.used", 
+"master.aliveWorkers", 
+"master.apps", 
+"master.waitingApps", 
+"master.workers"
+```
+
+## Proposed solution
+
+To implement a Graphite metrics solution with the current Spark deployment
+methodology, there will need to be changes to the oshinko-rest application,
+the addition of a configuration file to current Spark images, and the
+creation of two images for the Graphite and Carbon components.
+
+The oshinko-rest application will need to be modified so that the deployments
+for Spark master pods will include the new Graphite and Carbon images. An
+example of what this looks like in YAML template form can be seen in
+example 2. The deployment will also need a service created for the Graphite
+API server to allow access to the metrics data.
+
+An open question for the oshinko-rest implementation is whether the user
+should be have the ability to disable the deployment of metrics, or if this
+feature should always be deployed. For this specification, we should assume
+that the metrics server is always deployed.
```

Comment:
Note, I would think this is just a flag, and could be included as one of the cluster configs. I'm guessing we could make metrics optional pretty easily once we have configs.

https://github.com/radanalyticsio/oshinko-specs/pull/1